viam-cartographer/src/slam_service/slam_service.cc

```diff
@@ -186,6 +178,56 @@ ::grpc::Status SLAMServiceImpl::GetInternalState(
     return grpc::Status::OK;
 }
 
+int SLAMServiceImpl::GetPointCloudMapC(
+    viam_carto_get_point_cloud_map_response *response) {
+    std::string pointcloud_map;
+    // Write or grab the latest pointcloud map in form of a string
+    try {
+        std::shared_lock optimization_lock{optimization_shared_mutex,
+                                           std::defer_lock};
+        if (action_mode != ActionMode::LOCALIZING &&
+            optimization_lock.try_lock()) {
+            // We are able to lock the optimization_shared_mutex, which means
+            // that the optimization is not ongoing and we can grab the newest
+            // map
+            GetLatestSampledPointCloudMapString(pointcloud_map);
+            std::lock_guard<std::mutex> lk(viam_response_mutex);
+            latest_pointcloud_map = pointcloud_map;
+        } else {
+            // Either we are in localization mode or we couldn't lock the mutex
+            // which means the optimization process locked it and we need to use
+            // the backed up latest map
+            if (action_mode == ActionMode::LOCALIZING) {
+                LOG(INFO)
+                    << "In localization mode, using cached pointcloud map";
+            } else {
+                LOG(INFO)
+                    << "Optimization is occuring, using cached pointcloud map";
+            }
+            std::lock_guard<std::mutex> lk(viam_response_mutex);
+            pointcloud_map = latest_pointcloud_map;
+        }
+    } catch (std::runtime_error &e) {
+        LOG(ERROR) << "Stopping Cartographer: error encoding pointcloud: "
+                   << e.what();
```

Comment:
If we don't want to catch all `runtime_errors`, we can change the type of exception thrown just thought we would want to change as little as possible for now, but let me know if you think we should be throwing more specific errors

Reply:
If we needed to handle different runtime errors with different return codes we could look into creating a class per runtime error, however what you have here is appropriate for our current use case.